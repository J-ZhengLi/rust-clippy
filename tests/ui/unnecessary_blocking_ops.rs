#![feature(async_fn_in_trait)]
#![feature(async_closure)]
#![allow(incomplete_features)]
#![warn(clippy::unnecessary_blocking_ops)]
use std::thread::sleep;
use std::time::Duration;
use std::{fs, io};
use tokio::io as tokio_io;

mod totally_thread_safe {
    pub async fn sleep(_dur: std::time::Duration) {}
}

pub async fn async_fn() {
    sleep(Duration::from_secs(1));
    //~^ ERROR: blocking function call detected in an async body
    fs::remove_dir("").unwrap();
    //~^ ERROR: blocking function call detected in an async body
    fs::copy("", "_").unwrap();
    //~^ ERROR: blocking function call detected in an async body
    let _ = fs::canonicalize("");
    //~^ ERROR: blocking function call detected in an async body

    {
        fs::write("", "").unwrap();
        //~^ ERROR: blocking function call detected in an async body
        let _ = io::stdin();
    }
    let _stdout = io::stdout();
    let mut r: &[u8] = b"hello";
    let mut w: Vec<u8> = vec![];
    io::copy(&mut r, &mut w).unwrap();
    //~^ ERROR: blocking function call detected in an async body
}

pub async fn non_blocking() {
    totally_thread_safe::sleep(Duration::from_secs(1)).await; // don't lint, not blocking

    let mut r: &[u8] = b"hello";
    let mut w: Vec<u8> = vec![];
    tokio_io::copy(&mut r, &mut w).await; // don't lint, not blocking
}

trait AsyncTrait {
    async fn foo(&self);
}

struct SomeType(u8);
impl AsyncTrait for SomeType {
    async fn foo(&self) {
        sleep(Duration::from_secs(self.0 as _));
        //~^ ERROR: blocking function call detected in an async body
    }
}

fn do_something() {}

fn closures() {
    let _ = async || sleep(Duration::from_secs(1));
    //~^ ERROR: blocking function call detected in an async body
    let async_closure = async move |_a: i32| {
        let _ = 1;
        do_something();
        sleep(Duration::from_secs(1));
        //~^ ERROR: blocking function call detected in an async body
    };
    let non_async_closure = |_a: i32| {
        sleep(Duration::from_secs(1)); // don't lint, not async
        do_something();
    };
}

fn main() {}
