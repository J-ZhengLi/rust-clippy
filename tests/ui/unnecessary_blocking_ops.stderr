error: blocking function call detected in an async body
  --> tests/ui/unnecessary_blocking_ops.rs:14:5
   |
LL |     sleep(Duration::from_secs(1));
   |     ^^^^^
   |
   = note: `-D clippy::unnecessary-blocking-ops` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::unnecessary_blocking_ops)]`

error: blocking function call detected in an async body
  --> tests/ui/unnecessary_blocking_ops.rs:16:5
   |
LL |     fs::remove_dir("").unwrap();
   |     ^^^^^^^^^^^^^^

error: blocking function call detected in an async body
  --> tests/ui/unnecessary_blocking_ops.rs:18:5
   |
LL |     fs::copy("", "_").unwrap();
   |     ^^^^^^^^

error: blocking function call detected in an async body
  --> tests/ui/unnecessary_blocking_ops.rs:20:13
   |
LL |     let _ = fs::canonicalize("");
   |             ^^^^^^^^^^^^^^^^

error: blocking function call detected in an async body
  --> tests/ui/unnecessary_blocking_ops.rs:24:9
   |
LL |         fs::write("", "").unwrap();
   |         ^^^^^^^^^

error: blocking function call detected in an async body
  --> tests/ui/unnecessary_blocking_ops.rs:31:5
   |
LL |     io::copy(&mut r, &mut w).unwrap();
   |     ^^^^^^^^

error: blocking function call detected in an async body
  --> tests/ui/unnecessary_blocking_ops.rs:50:9
   |
LL |         sleep(Duration::from_secs(self.0 as _));
   |         ^^^^^

error: blocking function call detected in an async body
  --> tests/ui/unnecessary_blocking_ops.rs:58:22
   |
LL |     let _ = async || sleep(Duration::from_secs(1));
   |                      ^^^^^

error: blocking function call detected in an async body
  --> tests/ui/unnecessary_blocking_ops.rs:63:9
   |
LL |         sleep(Duration::from_secs(1));
   |         ^^^^^

error: blocking function call detected in an async body
  --> tests/ui/unnecessary_blocking_ops.rs:75:33
   |
LL |     std::thread::spawn(async || sleep(Duration::from_secs(1)));
   |                                 ^^^^^

error: aborting due to 10 previous errors

